### ISSUE TITLE
Some actual proper design for internals

### ISSUE THREAD

**Issue by maciejhirsz** [maciejhirsz/beef#15]
Messing around between the fat pointer and thin pointer is quite a lot of work, and there is quite some deep coupling between `generic::Cow`, the `Beef` trait and the `Capacity`. `Capacity` is also a poor name for what the trait is doing now.

A reasonable logical split would be:
+ `generic::Cow` implements the API surface, methods like `borrow`, `owned` and `into_owned`.
+ `Beef` sits in the middle and converts borrowed and owned values into parts (pointer, len, cap).
+ Rename `Capacity` to something like `Container`, have it store (pointer, len, cap) from `Beef` internally so that the particulars of how capacity is handled doesn't leak out to other abstractions.

**Comment by pickfire** [comment]
Is there a reason why it needs to be (pointer, len, cap) and not (pointer, cap, len)? Is it related to performance?

**Comment by maciejhirsz** [comment]
No, the `Container` here would be able to arrange stuff in whatever order it wants. Currently lean version is (ptr, len) with capacity put into higher 32 bits of len, so that obtaining a reference can be done by just masking those bits.

**Comment by pickfire** [comment]
Yeah, regarding the current design, I was asking why capacity were put into the higher 32 bits of length rather than length were put into the higher 32 bits of capacity? Isn't capacity being used more since it also stores the tag when capacity is 0?

**Comment by maciejhirsz** [comment]
No, capacity is only used:
+ When you can `into_owned`.
+ When the `Cow` is cloned.
+ When the `Cow` is dropped.

Most of what you want to do with a `Cow` is to read from it, which means creating a `&ref` (via `Deref` or `Borrow`). With the pointer and length in the same position for owned and borrowed values, getting a `&ref` is, unlike std implementation, branch free.
